### PR TITLE
WIP relations: SQLAlchemy approach

### PR DESCRIPTION
**Pull request by kaplun**
Signed-off-by: Samuele Kaplun samuele.kaplun@cern.ch

### CODE REVIEW

**Comment by kaplun** [comment]
@jacquerie @eamonnmag this is a prototype of table based relations. WDYT?

File: inspirehep/modules/relations/__init__.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    if len(path_parts) < 2:
+        current_app.logger.error('Bad JSONref URI: {0}'.format(uri))
+        return None
+
+    record_type = path_parts[-2]
+    recid = path_parts[-1]
+    return PersistentIdentifier.get(record_type, recid).id
+
+
+def crawl_json_for_ref(json, root='', relation=''):
```
Comment (kaplun) [inline inspirehep/modules/relations/__init__.py]: Recursively crawl a json looking for all the `$ref`s and keeping track of the corresponding JSON path (e.g. `authors[0]`) and relation (e.g. `authors`).

File: inspirehep/modules/relations/__init__.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+from flask import current_app
+
+from werkzeug.urls import url_parse
+from invenio_db import db
+from invenio_records.signals import after_record_insert, after_record_update
+
+from .models import Relations
+
+def uri_to_uuid(uri):
```
Comment (kaplun) [inline inspirehep/modules/relations/__init__.py]: This transform a URI from `$ref` in the corresponding `uuid` to be used in SQL.

File: inspirehep/modules/relations/__init__.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    ))
+    return ret
+
+
+@after_record_insert.connect
+@after_record_update.connect
+def upsert_relation(json):
+    with db.session.begin_nested():
+        id_from = uri_to_uuid(json['$self'])
+        Relations.query().filter(Relations.id_from == id_from).delete()
```
Comment (kaplun) [inline inspirehep/modules/relations/__init__.py]: First deletes all outgoing relations from this record and then re-upload them.

**Comment by kaplun** [comment]
The main thing missing with such approach Vs. the proposed Neo4J one is that I can not filter records via proporties that are not relations. 

E.g. how can I consider only those publication from a given year?

**Comment by jacquerie** [comment]
A similar design was proposed long ago by several people (among which me). It was a bad idea, and it still is: it is essentially reimplementing a graph database in a relational one.

**Comment by kaplun** [comment]
That ii think is what @eamonnmag was proposing yesterday. I mean at some point we have to give up on "this is bad because it implements feature X which is instead for tool Y" since this is incompatible with "we are adding service over service and this is bad".

Why is this approach bad? I'd it really a problem to reimplement? I guess there are predominance implications but are we hitting them at our scale? Compared we the benefit of being able to do joins directly in the db and have transactions consistency?

Alternatively the same technique could be used to populate Neo4j but there we will have to extend it with additional properties if we want to allow for further filtering.

**Comment by kaplun** [comment]
P.s. and including me https://github.com/inveniosoftware/invenio/issues/2573 :smile:

**Comment by jacquerie** [comment]
Postponed, it will be resurrected if necessary.

**Pull request closed without merging**